### ISSUE TITLE
Invalid TypeScript signatures for having, andHaving and orHaving

### ISSUE THREAD

**Issue by danielrearden** [knex/knex#3714]
# Environment

Knex version: 0.20.11
Database + version: any
OS: MacOS 10.14.16

@lorefnon 

# Bug

1. Explain what kind of behavior you are getting and how you think it should do

The TypeScript signatures for the having methods should match the source code and the docs. Instead, they contain an extra parameter, and the existing parameters are not typed correctly.

```
interface QueryInterface<TRecord extends {} = any, TResult = any> {
  ...
  having: Having<TRecord, TResult>;
  andHaving: Having<TRecord, TResult>;
  orHaving: Having<TRecord, TResult>;
}

interface Having<TRecord = any, TResult = unknown[]>
  extends RawQueryBuilder<TRecord, TResult>,
    WhereWrapped<TRecord, TResult> {
  <K1 extends keyof TRecord, K2 extends keyof TRecord>(
    tableName: string,
    column1: K1,
    operator: ComparisonOperator,
    column2: K2
  ): QueryBuilder<TRecord, TResult>;
  (
    tableName: string,
    column1: string,
    operator: string,
    column2: string
  ): QueryBuilder<TRecord, TResult>;
}
```

The signatures here indicate 4 parameters, where as [the docs](https://knexjs.org/#Builder-having) and the [source](https://github.com/knex/knex/blob/4feefdf3d2532c23f2f662c538bf2d81700daeb1/lib/query/builder.js#L691) show only three. The `tableName` parameter is extraneous -- it should just be `column1`, `operator` and `column2`. If you attempt to call the function with 4 parameters, you'll get an invalid operator error because it's assumed the 2nd parameter is the operator.

Additionally, instead of `column1` and `column2`, the parameters should just be named `column` and `value`, with the types resembling the `Where` definitions -- `column` could be either `string` or `Raw` and value could be `Value` or `null` or `QueryBuilder`. Something like:

```
interface Having<TRecord = any, TResult = unknown[]>
  extends RawQueryBuilder<TRecord, TResult>,
    WhereWrapped<TRecord, TResult> {
  (
  (columnName: string, operator: string, value: Value | null): QueryBuilder<TRecord, TResult>;
  (columnName: string, operator: string, value: QueryBuilder<...>): QueryBuilder<TRecord, TResult>;
  (left: Raw, operator: string, right: Value | null): QueryBuilder<TRecord, TResult>;
  (left: Raw, operator: string, right: QueryBuilder<...>): QueryBuilder<TRecord, TResult>;
}
``` 

2. Error message

> Argument of type 'Raw<any>' is not assignable to parameter of type 'string'.ts(2345)

3. Reduced test code

```
import Knex from 'knex'

const knex = Knex({
  client: 'pg',
  connection: {
    host: '127.0.0.1',
    user: 'your_database_user',
    password: 'your_database_password',
    database: 'myapp_test',
  },
})

const query = knex('foos').having(knex.raw('??', ['someColumn']), '=', 10)
```

**Comment by lorefnon** [comment]
Yes, you are right. Thanks for reporting this. 

This was not detected by our tests because the `Having` interface also (incorrectly) extends `RawQueryBuilder` and many common cases get (incorrectly) matched by this signature of `RawQueryBuilder` 

```
   (
      sql: string,
      ...bindings: readonly RawBinding[]
    ): QueryBuilder<TRecord, TResult2>
```

I'll take care of fixing this.

**Comment by danielrearden** [comment]
Out of curiosity, is there a reason the HAVING methods are different from the WHERE methods? I was under the impression that both clauses just had to return a boolean expression -- i.e. a valid WHERE expression would also be a valid HAVING expression and vice versa (although maybe this is dialect-specific?).

I imagine having the same API for both clauses would simplify both the code and the type definitions, not to mention make the HAVING methods more robust.

**Comment by lorefnon** [comment]
I think the only difference in signatures is that implicit equality is not supported in having. `.where('name', 'lorefnon')` compiles to `where name = 'lorefnon'` but in case of having the operator must be specified. 

I don't know the exact reason why the API would be like that but I can guess that in having clause we usually operate on aggregated values so the conditions are generally lt/gt/range comparisons as opposed to equality (which are more common for discreet values).

**Comment by kibertoad** [comment]
Released in 0.20.12
